### ISSUE TITLE
bench, pgo, release profile

### ISSUE THREAD

**Issue by mibere** [mhonert/velvet-chess#13]
Hi there,

are there plans to implement a 'bench' command, that can be used for [PGO](https://doc.rust-lang.org/rustc/profile-guided-optimization.html) during compilation and to compare different compile settings?

And, is there a benefit to set a [release profile](https://doc.rust-lang.org/book/ch14-01-release-profiles.html) in Cargo.toml?

```
[profile.release]
...
```

with settings like [opt-level](https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level), [lto](https://doc.rust-lang.org/cargo/reference/profiles.html#lto), [panic](https://nnethercote.github.io/perf-book/build-configuration.html#abort-on-panic), [strip](https://nnethercote.github.io/perf-book/build-configuration.html#strip-debug-info-and-symbols)

[Here](https://nnethercote.github.io/perf-book/build-configuration.html#summary) we read

> If you want to maximize runtime speed, consider all of the following: codegen-units = 1, lto = "fat", an alternative allocator, and panic = "abort".

Example

```
[profile.release]
codegen-units = 1
opt-level = 3
lto = = "fat"
panic = "abort"
strip = true
```

Thanks for your engine, your time and work. Keep on!

**Comment by mhonert** [comment]
Hi,

thanks for the suggestions.
A while ago, I compared the performance impact of several options and also had a look at PGO.
Back then, my conclusion was to leave most parameters to the defaults, but only set "codegen-units = 1" in the Cargo.toml (opt-level 3 is already the default for release builds).
It increased compile times significantly, but also produced the most performant code (at least on the 3 CPUs I tested).

But in the meantime, new Rust and LLVM versions have been released and it might make sense to check the best parameters again.
I did a short test today with the default setting for codegen-units (which is 16 for release builds) and with lto set to "thin" and it was slightly faster.

I will do some more comparisons and also implement a bench command for the next release.

Best regards.
